frontend/src/features/location/locationMiddleware.ts

```diff
@@ -0,0 +1,99 @@
+import { createListenerMiddleware, createAction } from "@reduxjs/toolkit";
+import * as Location from "expo-location";
+import { useEffect } from "react";
+
+import { useAppDispatch } from "../../common/hooks";
+
+import { updateLocationStatus } from "./locationSlice";
+
+const locationMiddleware = createListenerMiddleware();
+const subscribeLocation = createAction<undefined>("location/start");
+const unsubscribeLocation = createAction<undefined>("location/stop");
+
+// This seems to be the only accuracy that's at least somewhat reliable for our purposes. Takes a few seconds
+// to initialize and then updates every second or so.
+const ACCURACY = Location.Accuracy.BestForNavigation;
+
+let locationSubscription: Location.LocationSubscription | null = null;
+
+// The problem with location updates is that they are a callback-based API with sensitive lifecycle limitations
+// that we need to cram into the redux API. This middleware attempts to abstract this process with the following
+// lifecycle:
+// 1. The top-level Main component calls manageLocationMiddleware(), which then sends a subscribe action to
+// the middleware. This creates the subscription and forwards them to the state in the location slice.
+// 2. The top-level Main component unmounts, which sends an unsubscribe action to the middleware. The middleware
+// immediately unsubscribes and frees the subscription. This way, we can stop tracking location as soon as the
+// app is no longer visible.
+
+locationMiddleware.startListening({
+  // actionCreator specifies the action that when dispatched will trigger the
+  // code below.
+  actionCreator: subscribeLocation,
+  effect: async (_, listenerApi) => {
+    if (locationSubscription !== null) {
+      // Nothing to do
+      return;
+    }
+
+    const { status } = await Location.requestForegroundPermissionsAsync();
+    if (status !== "granted") {
+      // Permission was not granted, we can't do anything. Send the outcome
+      // to the companion slice state.
+      listenerApi.dispatch(updateLocationStatus({ type: "not_granted" }));
+      return;
+    }
+
+    try {
+      // Have to track the current subscription so we can unsubscribe later.
+      locationSubscription = await Location.watchPositionAsync(
+        { accuracy: ACCURACY },
+        (newLocation) => {
+          // Forward updates to the companion slice so that components can
+          // use the current state.
+          listenerApi.dispatch(
+            updateLocationStatus({
+              type: "active",
+              location: newLocation.coords,
+            }),
+          );
+        },
+      );
+    } catch (e) {
+      listenerApi.dispatch(updateLocationStatus({ type: "error" }));
+    }
+  },
+});
+
+locationMiddleware.startListening({
```

Comment:
can you add some comments here explaining why we have two .startListening functions? I'm starting to understand but this is still a bit confusing to me.

Reply:
ok from discussion with GPT it seems like we are creating two listeners? One for subscribeLocation and one for unsubscribeLocation? I would just add some comments before each